### PR TITLE
GDB-10569: WB Migration: implement better deep copy functionality

### PR DESCRIPTION
**Pull request by boyan-tonchev**
## What
Implement deep copy functionality.

## Why
We use subscriptions to handle changes to certain values. When a context property value is changed, all subscribers are notified of the change and retrieve the new value. To ensure that each subscriber works with an independent copy of the value, we use deep copies. This guarantees that there are no side effects when one subscriber modifies the value. If a value needs to be changed, the client modifies its copy and updates it through the corresponding model's context service.

## How
Create an abstract model class that implements the Copyable interface. This class will provide a copy method. All application models must extend this base model class. The class will also include generic methods common to all models. Create an abstract ModelList class for managing lists of models. This class will include generic methods for working with arrays of models.

## Checklist
- [X] Branch name
- [X] Target branch
- [X] Commit messages
- [X] Squash commits
- [X] MR name
- [X] MR Description

### CODE REVIEW

File: packages/api/src/services/context.servce.spec.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+    // Given the property is updated before the callback function is registered
+    contextService.updateProperty(propertyName, valueOfTestProperty);
+
+    // When a callback function is subscribed
+    const callBackFunction = jest.fn();
+    contextService.subscribeToProperty(propertyName, callBackFunction);
+
+    // Then the callback function should be called with the last value
+    expect(callBackFunction.mock.lastCall[0]).toEqual(valueOfTestProperty);
```
Comment (plamen-yordanov) [inline packages/api/src/services/context.servce.spec.ts:37]: `expect(callBackFunction).toHaveBeenCalledWith(valueOfTestProperty);`  or 
`expect(callBackFunction).toHaveBeenLastCalledWith(valueOfTestProperty);` depending on what you need, will be a bit more readable here

File: packages/api/src/services/context.servce.spec.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+class TestContextService extends ContextService {
+  public updateProperty<T>(propertyName: string, value: T): void {
+    this.updateContextProperty(propertyName, value);
+  }
+
+  public getProperty<T>(propertyName: string): T | undefined {
+    return this.getContextPropertyValue(propertyName);
+  }
+
+  public subscribeToProperty<T>(propertyName: string, callback: (value: T | undefined) => void): () => void {
```
Comment (plamen-yordanov) [inline packages/api/src/services/context.servce.spec.ts:13]: `(value: T | undefined)` can be shortened to `(value?: T)`

File: packages/api/src/services/context.servce.spec.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+    // Given two subscribed functions for 'testProperty' changes
+    contextService.subscribeToProperty(propertyName, firstCallBackFunction);
+    contextService.subscribeToProperty(propertyName, secondCallBackFunction);
+
+    // When the property is updated
+    contextService.updateProperty(propertyName, valueOfTestProperty);
+
+    // Then both callback functions should be called with the value of the property
+    expect(firstCallBackFunction.mock.lastCall[0]).toEqual(valueOfTestProperty);
```
Comment (plamen-yordanov) [inline packages/api/src/services/context.servce.spec.ts:93]: Same as the previous comment

File: packages/api/src/services/language/language-context.service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    // When I register a callback function for language changes,
+    languageContextService.onSelectedLanguageChanged(onSelectLanguageCallbackFunction);
+    // and the selected language is changed.
+    languageContextService.updateSelectedLanguage(newLanguage);
+
+    // Then I expect the callback function to be called with the passed language.
+    expect(onSelectLanguageCallbackFunction.mock.lastCall[0]).toEqual(newLanguage);
+  });
+
+  test('should stop calling the "onSelectLanguageCallbackFunction" when unsubscribed from the event.', () => {
```
Comment (svilenvelikov) [inline packages/api/src/services/language/language-context.service.spec.ts:23]: Remove this test. It duplicates whats already tested with the ContextService

File: packages/api/src/services/repository-location/repository-location-context.service.spec.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    // When I register a callback function for active location changes,
+    repositoryLocationContextService.onActiveLocationChanged(onActiveLocationChangedCallbackFunction);
+    // and the active repository location is updated,
+    repositoryLocationContextService.updateActiveLocation(newRepositoryLocation);
+
+    // Then I expect the callback function to be called with the passed repository location.
+    expect(onActiveLocationChangedCallbackFunction.mock.lastCall[0]).toEqual(newRepositoryLocation);
+  });
+
+  test('Should stop calling the "onActiveLocationChangedCallbackFunction" when unsubscribed from the event.', () => {
```
Comment (svilenvelikov) [inline packages/api/src/services/repository-location/repository-location-context.service.spec.ts:24]: Duplication. Remove this test!

File: packages/api/src/services/repository-location/repository-location.service.spec.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import {ServiceProvider} from '../../service.provider';
+import {RepositoryLocationService} from './repository-location.service';
+import {RepositoryLocationRestService} from './repository-location-rest.service';
+import {RepositoryLocation} from '../../models/repository-location';
+
+// Mock the dependencies
```
Comment (svilenvelikov) [inline packages/api/src/services/repository-location/repository-location.service.spec.ts:6]: I don't know if you are writing these comments or the AI, but they are a bit excessive. Please, limit them to the places where stuff really needs commenting.

File: packages/api/src/services/repository/repository-context.service.spec.ts [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+      // and the selected repository is changed,
+      repositoryContextService.updateSelectedRepository(newRepository);
+
+      // Then I expect the callback function to be called with the passed repository,
+      expect(onSelectedRepositoryChangedCallBackFunction.mock.lastCall[0]).toEqual(newRepository);
+      // but the repository should be a different instance.
+      expect(onSelectedRepositoryChangedCallBackFunction.mock.lastCall[0]).not.toBe(newRepository);
+    });
+
+    test('Should stop calling the "onSelectedRepositoryChangedCallbackFunction" when unsubscribed from the event.', () => {
```
Comment (svilenvelikov) [inline packages/api/src/services/repository/repository-context.service.spec.ts:28]: remove it

File: packages/api/src/services/repository/repository-context.service.spec.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+      // and the repositories are updated,
+      repositoryContextService.updateRepositories(newRepositories);
+
+      // Then I expect the callback function to be called with the passed repositories,
+      expect(onRepositoriesChangedCallbackFunction.mock.lastCall[0]).toEqual(newRepositories);
+      // but the repositories should be a different instance.
+      expect(onRepositoriesChangedCallbackFunction.mock.lastCall[0]).not.toBe(newRepositories);
+    });
+
+    test('Should stop calling the "onRepositoriesChangedCallbackFunction" when unsubscribed from the event.', () => {
```
Comment (svilenvelikov) [inline packages/api/src/services/repository/repository-context.service.spec.ts:64]: remove it and them remove the nested describe blocks

File: packages/api/src/services/repository/repository.service.spec.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+    ServiceProvider.get = jest.fn().mockReturnValue(mockRepositoryRestService);
+    // Initialize RepositoryService.
+    repositoryService = new RepositoryService();
+  });
+
+  describe('getRepositories', () => {
+    test('should return repositories', async () => {
+      // Mock the response from RepositoryRestService.
+      const rawRepository = {id: 'repo-id', title: 'repo-title', location: ''};
+      const mockResponse = {"": [rawRepository]};
```
Comment (plamen-yordanov) [inline packages/api/src/services/repository/repository.service.spec.ts:26]: shouldn't these be single quotes?

File: packages/api/src/services/context.servce.spec.ts [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+  public getProperty<T>(propertyName: string): T | undefined {
+    return this.getContextPropertyValue(propertyName);
+  }
+
+  public subscribeToProperty<T>(propertyName: string, callback: (value: T | undefined) => void): () => void {
+    return this.subscribe(propertyName, callback);
+  }
+}
+
+describe("ContextService", () => {
```
Comment (plamen-yordanov) [inline packages/api/src/services/context.servce.spec.ts:18]: I assume these should be single quotes also?

File: packages/api/src/services/context.servce.spec.ts (hunk 1 above)
Comment (boyan-tonchev) [reply packages/api/src/services/context.servce.spec.ts]: Done

File: packages/api/src/services/context.servce.spec.ts (hunk 2 above)
Comment (boyan-tonchev) [reply packages/api/src/services/context.servce.spec.ts]: Done

File: packages/api/src/services/context.servce.spec.ts (hunk 3 above)
Comment (boyan-tonchev) [reply packages/api/src/services/context.servce.spec.ts]: Done

File: packages/api/src/services/language/language-context.service.spec.ts (hunk 4 above)
Comment (boyan-tonchev) [reply packages/api/src/services/language/language-context.service.spec.ts]: Removed

File: packages/api/src/services/repository-location/repository-location-context.service.spec.ts (hunk 5 above)
Comment (boyan-tonchev) [reply packages/api/src/services/repository-location/repository-location-context.service.spec.ts]: Removed

File: packages/api/src/services/repository-location/repository-location.service.spec.ts (hunk 6 above)
Comment (boyan-tonchev) [reply packages/api/src/services/repository-location/repository-location.service.spec.ts]: Removed

File: packages/api/src/services/repository/repository-context.service.spec.ts (hunk 7 above)
Comment (boyan-tonchev) [reply packages/api/src/services/repository/repository-context.service.spec.ts]: Removed

File: packages/api/src/services/repository/repository-context.service.spec.ts (hunk 8 above)
Comment (boyan-tonchev) [reply packages/api/src/services/repository/repository-context.service.spec.ts]: Removed

File: packages/api/src/services/repository/repository.service.spec.ts (hunk 9 above)
Comment (boyan-tonchev) [reply packages/api/src/services/repository/repository.service.spec.ts]: Done

File: packages/api/src/services/context.servce.spec.ts (hunk 10 above)
Comment (boyan-tonchev) [reply packages/api/src/services/context.servce.spec.ts]: Done

**Review by svilenvelikov — APPROVED** [review]

**Pull request merged**